Move excluded files to config array

sharidas opened the pull request:
Moving excluded files to config array.
Instead of populating in the ExcludeFile* 
php files, we can have it in config.php.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->
This change would help users to mention folders/files to be excluded during integrity check to the config.php. In the config.php it will be represented as an array.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://github.com/owncloud/core/issues/23186
https://github.com/owncloud/enterprise/issues/1874

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Instead of hard coding the files/folders in the files, its better to move them to the config.php. Also this change would help the integrity check to exclude files/folders mentioned in the config.php

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

tomneedham commented on lib/private/IntegrityCheck/Checker.php:
```diff
@@ -398,13 +398,21 @@ private function verify($signaturePath, $basePath, $certificateCN, $force = fals
 		}
 
+		//Exclude files which shouldn't fall for comparison
+		$excludeFiles = \OC::$server->getSystemConfig()->getValue('excludedFiles', '');
```
default should be an array since we use it in `in_array` later

tomneedham commented:
should also include a sample entry in config.sample.php

DeepDiver1975 commented on config/config.sample.php:
```diff
@@ -1205,6 +1205,37 @@
+'excludedFolders' =>
+	array (
+		0 => '/home/sujith/test/owncloud/data',
+		1 => '/home/sujith/test/owncloud/themes',
```
better use OC::$SERVERROOT

example
```php
OC::$SERVERROOT . '/assets'

DeepDiver1975 replied:
and we shall find good example folders - maybe two are enough.
/apps/my-theme might be a good example and /assets

DeepDiver1975 commented on config/config.sample.php:
```diff
@@ -1205,6 +1205,37 @@
+/**
+ * Exclude file name patterns from the integrity checker command
+ */
+'excludedFilenamePatterns' =>
```
isn't the pattern enough to also properly exclude folders and files from above?

PVince81 replied:
not sure... maybe some files could have prefixes/suffixes ? I'd also prefer the simpler solution.

@sharidas Did anyone report such cases in the original ticket ?

sharidas replied:
@PVince81 No I didn't find any cases related to exclude filename patterns in the original ticket. I think we can stick on with exclude file names and remove exclude filename patterns from the PR.

DeepDiver1975 commented on lib/private/IntegrityCheck/Checker.php:
```diff
@@ -398,13 +398,21 @@ private function verify($signaturePath, $basePath, $certificateCN, $force = fals
 		}
 
+		//Exclude files which shouldn't fall for comparison
+		$excludeFiles = \OC::$server->getSystemConfig()->getValue('excludedFiles', []);
```
inject config instance - no use of \OC::$server inside any class please

DeepDiver1975 commented on lib/private/IntegrityCheck/Iterator/ExcludeFileByNameFilterIterator.php:
```diff
@@ -78,4 +85,12 @@ public function accept() {
 	}
+
+	protected function getExcludedFileNames() {
+		return \OC::$server->getConfig()->getSystemValue('excludedFiles', '');
```
keep proper defaults - basically just like before

DeepDiver1975 commented on lib/private/Setup.php:
```diff
@@ -317,6 +340,9 @@ public function install($options) {
 			'overwrite.cli.url'	=> $request->getServerProtocol() . '://' . $request->getInsecureServerHost() . \OC::$WEBROOT,
 			'dbtype'			=> $dbType,
 			'version'			=> implode('.', \OCP\Util::getVersion()),
+			'excludedFolders'   => $excludeFolders,
```
I would no set them during setup. the defaults will be in our code base and people can extend/change the list if they like.

DeepDiver1975 commented on tests/lib/IntegrityCheck/CheckerTest.php:
```diff
@@ -497,114 +497,6 @@ public function testWriteCoreSignature() {
 		$this->checker->writeCoreSignature($x509, $rsa, \OC::$SERVERROOT . '/tests/data/integritycheck/app/');
 	}
 
-	public function testWriteCoreSignatureWithUnmodifiedHtaccess() {
```
where did these tests go?

PVince81 commented:
I heard from @Peter-Prochaska that it might not be a good idea to allow excluding whole folders, so not sure... See https://github.com/owncloud/core/issues/29094#issuecomment-335396910

PVince81 commented:
- [ ] Please only allow excluding files, not folders, as discussed with @Peter-Prochaska

mmattel commented:
Maybe only a small detail to avoid naming confusion:
``excludedFolders`` --> ``integrityExcludedFolders`` also valid for the others.

Reason: https://doc.owncloud.org/server/latest/admin_manual/configuration/server/excluded_blacklisted_files.html

sharidas committed:
Move excluded files to config array

Moving excluded files to config array.
Instead of populating in the ExcludeFile*
php files, we can have it in config.php.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented:
Updated this PR to exclude files only and not folders.

mmattel commented on config/config.sample.php:
```diff
@@ -1210,6 +1210,25 @@
+ * Exclude directories from the integrity checker command
+ */
+'excludedFiles' =>
+	array (
```
Description text: Exclude directories --> Exclude files

mmattel replied:
Also see my comment above to avoid confusion:
This variable should be renamed from ``excludedFiles`` to ``integrityExcludedFiles``

sharidas commented:
Not sure how to make codecov happy for config.sample.php. Any pointers would be helpful. Thanks.

PVince81 commented on config/config.sample.php:
```diff
@@ -1210,6 +1210,25 @@
+ */
+'integrityExcludedFiles' =>
+	array (
+		0 => '.DS_Store',
```
no need for the numeric index

sharidas replied:
Updated the config.sample.php by removing numeric index.

sharidas commented:
Backport PR to stable10 branch: https://github.com/owncloud/core/pull/29460

PVince81 commented on config/config.sample.php:
```diff
@@ -1210,6 +1210,25 @@
 /**
+ * Exclude directories from the integrity checker command
+ */
+'integrityExcludedFiles' =>
```
why camel case ? all of the other config entries are either dot separated, underscore separated or have no separator for all-lowercase. The latest trend is to use dots as separators.

sharidas replied:
Updated to `integrity.excluded.files`

PVince81 commented:
@Peter-Prochaska second review ?

Peter-Prochaska commented on lib/private/IntegrityCheck/Checker.php:
```diff
@@ -230,7 +230,7 @@ private function generateHashes(\RecursiveIteratorIterator $iterator,
 	 * @param X509 $certificate
 	 * @param RSA $privateKey
-	 * @return string
+	 * @return mixed
```
Why mixed? it's an array

sharidas replied:
@Peter-Prochaska  Updated the PR.

mmattel commented on config/config.sample.php:
```diff
@@ -1210,6 +1210,18 @@
 /**
+ * Exclude directories from the integrity checker command
+ */
+'integrity.excluded.files' =>
```
@PVince81 we still have the description text mismatch stating that we exclude directories while we exclude files ...
This inconsistancy needs to be fixed and backported

mmattel replied:
@PVince81 just seen that is has been fixed ...
